### DISCUSSION TITLE
Getting a 400 error when trying to authenticate.

### DISCUSSION THREAD

**Discussion by kapilkaisare** [tidal-music/discussions#54] [Q&A]
I'm working with the Developer API, trying to connect it to ChatGPT, and am unable to authenticate with the service. I have the following values:

1. Client Id
2. Client Secret
3. Authorization values (https://login.tidal.com/authorize)
4. Token Url (https://auth.tidal.com/v1/oauth2/token)
5. Scopes: collection.read, collection.write, playlist.read, playlist.write
6. A callback url: (https://chat.openai.com/aip/a-really-long-string/oauth/callback)

When ChatGPT invokes authentication, I see the following call

`
https://login.tidal.com/authorize?response_type=code&client_id=<client-id>&redirect_uri=https%3A%2F%2Fchat.openai.com%2Faip%2Fa-really-long-string%2Foauth%2Fcallback&state=b890b4a4-b804-439d-88fa-b942b10b706d&scope=collection.read%2C+collection.write%2C+playlist.read%2C+playlist.write`

clear with a 302 (found). This call is followed by another to:

`https://login.tidal.com/authorize?client_id=<client-id>&redirect_uri=https%3A%2F%2Fchat.openai.com%2Faip%2Fa-really-long-string%2Foauth%2Fcallback&response_type=code&scope=collection.read%2C%20collection.write%2C%20playlist.read%2C%20playlist.write&state=another-really-long-string&geo=US&campaignId=default`

which clears with a 200. I am eventually redirected to 

`https://login.tidal.com/api/refreshlogin?client_id=<client-id>&redirect_uri=https%3A%2F%2Fchat.openai.com%2Faip%2Fa-really-long-string%2Foauth%2Fcallback&response_type=code&scope=collection.read,%20collection.write,%20playlist.read,%20playlist.write&state=another-reaaly-long-string&geo=US&campaignId=default`

which renders the following screen. 

![image](https://github.com/tidal-music/discussions/assets/3334563/28d0f30a-9105-4a3f-bd7b-f62d5fb27d12)

I'm already logged in, and when I click 'Yes, continue', the page gets rerendered with the same error. code 400. The response is :

```
hasInternalDetails: false
isApiError: true
isAxiosError: true
message: "Request failed with status code 400"
messageCode: "LOGIN_AGAIN"
status: 400
```

If I try to switch users, and try to sign in with the same account, I see a similar error, and I observe that a call to the url

`https://login.tidal.com/api/email/user/existing?client_id=cYFAriCVGb79jLNZ&default_config=true&id=12036&name=Open%20API%20%28devId%3A%20127a0e58-d495-45f5-a566-7f4b9237ff6a,%20appCreated%3A%201712967925568%29&redirect_uri=https%3A%2F%2Fchat.openai.com%2Faip%2Fg-08e2ec3cfac9861e05997f9e78c46eed4b80eb34%2Foauth%2Fcallback&response_type=code&scope=collection.read,%20collection.write,%20playlist.read,%20playlist.write&state=b890b4a4-b804-439d-88fa-b942b10b706d&geo=US&campaignId=default`

has failed with the same 400 code and the following response:

```
hasInternalDetails: false
isApiError: true
isAxiosError: true
message: "Request failed with status code 400"
messageCode: "GENERIC_ERROR"
status: 400
```

Does anyone else see what I might be doing wrong?

**Comment by stefanmoro** [comment]
Hey @kapilkaisare ,

I believe that your issue might be caused by you trying to acquire a non-existing scope. TIDAL Developer does not yet allow access to user data such as playlists (we are working hard to enable it though), and some of the scopes you requested does not exist.

**Reply by kapilkaisare** [reply]
Hi @stefanmoro ,

Thanks for the response! I modified the scopes to `collection.read, collection.write`, but am still seeing the same error.

**Reply by kapilkaisare** [reply]
Hi @stefanmoro,

Any idea what I might be doing wrong? When I try authenticating, I see the following call

```
https://login.tidal.com/authorize?response_type=code&client_id=BoeBO5MAQ2chZoTp&redirect_uri=https%3A%2F%2Fchat.openai.com%2Faip%2Fg-e0eb9a7c6e52907006abe20f07695e97b6ffba2e%2Foauth%2Fcallback&state=df069282-aa0e-4baf-b622-f84637ae5374&scope=collection.read%2C+collection.write%2C+playlists.read%2C+playlists.write
``` 

go through with a 302. Note that the scopes match the options that I see in the dashboard. This is followed by a call to 

```
https://login.tidal.com/authorize?client_id=BoeBO5MAQ2chZoTp&redirect_uri=https%3A%2F%2Fchat.openai.com%2Faip%2Fg-e0eb9a7c6e52907006abe20f07695e97b6ffba2e%2Foauth%2Fcallback&response_type=code&scope=collection.read%2C%20collection.write%2C%20playlists.read%2C%20playlists.write&state=df069282-aa0e-4baf-b622-f84637ae5374&geo=US&campaignId=default
```

which clears with a 200. However, the next call in the flow fails with a 400:

```
https://login.tidal.com/api/refreshlogin?client_id=BoeBO5MAQ2chZoTp&redirect_uri=https%3A%2F%2Fchat.openai.com%2Faip%2Fg-e0eb9a7c6e52907006abe20f07695e97b6ffba2e%2Foauth%2Fcallback&response_type=code&scope=collection.read,%20collection.write,%20playlists.read,%20playlists.write&state=df069282-aa0e-4baf-b622-f84637ae5374&geo=US&campaignId=default
```

**Reply by eetuhuhtamaki** [reply]
Hi, I'm having similar problems. Did you ever figure this out? Thank you in advance.

**Reply by stefanmoro** [reply]
I'd recommend checking out our [API reference](https://developer.tidal.com/reference/web-api), where it is also possible to login, acquire scopes and try out the API. That is where you'll find the most up to date information about the API. Also, check out the [Authorization](https://developer.tidal.com/documentation/api-sdk/api-sdk-authorization) documentation.

Hope that helps.
